### ISSUE TITLE
[$150] Jump to post

### ISSUE THREAD

**Issue by gondzo** [appirio-tech/connect-app#2306]
We need a way to jump to specific topic or post based only on topic or post ID. This would require fetching the data backwards - fetch post and the topic from the api, figure out the reference codes and load the correct project and correct tab (plan or dashboard) with the referenced post displayed correctly (phase posts expanded and loaded to correct date/time).
IMO this would best be achieved with a new route like connect.topcoder.com/posts/postId and connect.topcoder.com/topics/topicId. Loading these routes would redirect the user to the correct route (project dashboard, plan, etc).
@vikasrohit any other ideas?

Aha! Link: https://topcoder.aha.io/features/TCCONNECT-294

**Comment by vikasrohit** [comment]
@gondzo we have this functionality built in, please refer [this](https://github.com/appirio-tech/connect-app/blob/dev/src/routes.jsx#L136). I think what we need is to update the logic to handle the phase topics and add an anchor tag to the new Post card.

**Comment by gondzo** [comment]
Yup, we need to update RedirectToProject with new routes for phase topics (we don't have deep links for phases - it should switch to plan tab, expand the phase and load posts tab) and some logic to load the exact post in the topic if we want to link to exact postId

**Comment by vikasrohit** [comment]
Agreed. Marking this issue for next release as we can not get it done for this release.

**Comment by narekcat** [comment]
@gondzo I can't access Aha! Link. Can you give me access or send invitation email ?

**Comment by narekcat** [comment]
I have worked on this issue today, but it's late and I need to have a rest. I'll finish this issue tomorrow.

**Comment by vikasrohit** [comment]
@narekcat there is no need to access the aha because that does not have any additional details.

**Comment by narekcat** [comment]
@gondzo @vikasrohit  I can get project's topic by it's id, but I can't get project's post or topic by post id and in messages api there aren't any endpoint for getting project's topic or post by post id. Are there any endpoint for getting that data by post id ?
I examine discussions and found out that if we go /projects/:projectId/discussions/:discussionId, it doesn't shows discussion, it says "Invalid date". I think there are some bug or feature is incomplete.

**Comment by narekcat** [comment]
@gondzo @vikasrohit Can project's phase has more then one topics ? If it can't that I can redirect to /projects/:projectId/plan/phases/:phaseId/topics when user goes to /projects/:projectId/topics/:topics and /projects/:projectId/plan/phases/:phaseId/posts/postId when user goes to /projects/:projectId/posts/:postId

**Comment by gondzo** [comment]
Phase will always have just one topic.
When user goes to /discussions/topicId this should be the logic to get the relevant topic
call messages_api/topics/topicId and check the tag field in the response:
- 'PRIMARY' - load /projects/:referenceId#feed-{:topicId} 
- 'MESSAGES' - load /projects/:referenceId/discussions/:topicId
- 'phase#:phaseId' - load /projects/:referenceId/plan, expand the correct phase and load the posts tab (this can be done either by adding phase/tab details into the route or by passing the parameters explicitly)

Same thing should be done for navigating to /discussions/posts/postId except that first we need to get the topic id by calling messages_api/posts/postId

**Comment by narekcat** [comment]
@gondzo Ok, thank you very much.

**Comment by narekcat** [comment]
@gondzo Can you give me exact endpoint for getting post by it's id ? Like in case of topic is "/topics/${topicId}/read/db". I tried "/posts/${postId}", but it returns status code 404.

**Comment by vikasrohit** [comment]
It should be `/v5/topics/{topicId}/posts/{postId}/read/db`.

**Comment by narekcat** [comment]
@vikasrohit Ok, but then it can't jump to post having only postId, in case of connect.topcoder.com/posts/postId.

**Comment by vikasrohit** [comment]
I don't think we need to handle `connect.topcoder.com/posts/postId`. For every notification (which is the originator for the user to navigate to the specific topic or post) we should have both `topicId` and `postId` so we should be able to navigate to a particular post by constructing url with both params and I think we are already doing something similar to load the specific topic or post on dashboard via `/projects/7210#feed-27775` for loading topic and `/projects/7210#comment-80286` for loading particular post.
So, in summary, you should expect both `topicId` and `postId` in URL and `postId` can be present in `#` of the URL to navigate to the specific post.

**Comment by narekcat** [comment]
@vikasrohit For this issue we need new functionality to jump to topic from `/topics/topicId` or jump to post from `/topics/topicId/posts/postId`. Am I right ?
If you copy past `/projects/7210#feed-27775` or `/projects/7210#comment-80286` in browser address line, you will not see topic or feed. It working only when all content is loaded and you clicking on post date.
And gonzo said https://github.com/appirio-tech/connect-app/issues/2306#issuecomment-410986418.

**Comment by narekcat** [comment]
@gondzo @vikasrohit I have fixed this issue. If user goes to `/topics/:topicId` then it will redirect to `/projects/:projectId/plan/phases/:phaseId/topics`, opens Project Plan tab, opens project phase in the middle of screen, opens topics tab and loads posts. Same will happen when user goes to `/topics/:topicId/posts/:postId`, but it loads all posts and shows post with postId in the middle of screen.
If there are some misunderstandings let me know.

**Comment by vikasrohit** [comment]
> If you copy past /projects/7210#feed-27775 or /projects/7210#comment-80286 in browser address line, you will not see topic or feed

It only works for the dashboard topics/posts. I have tested this recently and it is working as expected. purpose of this issue is to apply the same logic for phase topics now.

> If there are some misunderstandings let me know.

Seems like you are now on correct track. One thing that I want to improve is the original URL which we are using load the specific phase topics/post because `/projects/:projectId/plan/phases/:phaseId/topics` and `/projects/:projectId/plan/phases/:phaseId/topics/:topicId/posts/:postId` requires new routes to be created and we have seen some issues in past because of new specific routes for phases, e.g. it makes the UI very jumpy when user clicks on a particular phase card. So, to avoid that I was thinking of using only `#` based approach as we are using for loading non phase topics and posts. May be something like `projects/:projectId/plan#feed-<feedId>` and `projects/:projectId/plan#comment-<feedId>-<commentId>` for loading specific topic and post for a phase. Once we have `feedId`, we can fetch the topic, and every topic is associated with a phase via `tag` field being `phase#<phaseId>`. So, now we have the phase id as well to open the the phase card.

**Comment by narekcat** [comment]
@vikasrohit Ok, I understand. Thank you.

**Comment by narekcat** [comment]
@vikasrohit @gondzo I have finished issue. You can use `projects/:projectId/plan#feed-<feedId>` and `projects/:projectId/plan#comment-<commentId>`, without feedId, for loading specific topic and post for a phase.

**Comment by gondzo** [comment]
@vikasrohit once minor issues are fixed, we'll have the two above routes for accessing the messages via direct link. 
What is left is updating /discussions route to redirect properly. Can we do that in the next release?

**Comment by vikasrohit** [comment]
> What is left is updating /discussions route to redirect properly. Can we do that in the next release?

Yes, no problem. Please create a separate issue to track that. Do you remember from where we are generating the `/discussions` link now?

**Comment by gondzo** [comment]
I don't think we're using /discussions link anywhere, it was in discourse emails only afaik. It will be useful to link to posts in email templates though

**Comment by vikasrohit** [comment]
Ohk. I think we are already using the same `#` based deep links to the messages in email templates.

**Comment by narekcat** [comment]
@gondzo @vikasrohit Are there any work to do with this issue ?

**Comment by gondzo** [comment]
@narekcat no, it's all done, we'll handle the remaining task in #2464

**Comment by narekcat** [comment]
@gondzo Ok, thank you.

**Comment by narekcat** [comment]
@gondzo @vikasrohit Sorry, but you forget to accepts this issue, and I don't see any payment for this issue in my profile.

**Comment by gondzo** [comment]
Payment has been processed - it takes a few hours for it to show up in your account

**Comment by narekcat** [comment]
@gondzo Ok, sorry, thank you very much.
